File: pkg/leapp/application.go
package leapp

import (
	"time"

	"github.com/cloudwindy/mirai/pkg/lue"
	"github.com/gofiber/fiber/v2"
	"github.com/gofiber/fiber/v2/middleware/session"
	"github.com/pkg/errors"
	lua "github.com/yuin/gopher-lua"
)

// Constants
const (
	methodAll     = "ALL"
	LTApplication = "Application"
)

type (
	StartHandler  func(listen string) error
	ReloadHandler func() error
	StopHandler   func(timeout time.Duration) error
)

type Config struct {
	App    fiber.Router
	Store  *session.Store
	Start  StartHandler
	Reload ReloadHandler
	Stop   StopHandler
}

type Application struct {
	c   Config
	up  bool
	sub bool
	fiber.Router
}

// New creates a new instance of the Lua engine factory.
func New(c Config) lue.Module {
	return func(E *lue.Engine) lua.LValue {
		// Create a new Fiber app
		app := new(Application)
		app.Router = c.App
		app.c = c

		// Set up the app functions
		index := E.NewTable()
		E.SetFuncs(index, appExports)

		return E.Class(LTApplication, app, index)
	}
}

var appExports = map[string]lue.Fun{
	"start":   appStart,
	"reload":  appReload,
	"stop":    appStop,
	"sub":     appSub,
	"use":     appUse,
	"add":     appAdd,
	"all":     appAddMethod(methodAll),
	"get":     appAddMethod(fiber.MethodGet),
	"head":    appAddMethod(fiber.MethodHead),
	"post":    appAddMethod(fiber.MethodPost),
	"put":     appAddMethod(fiber.MethodPut),
	"delete":  appAddMethod(fiber.MethodDelete),
	"connect": appAddMethod(fiber.MethodConnect),
	"options": appAddMethod(fiber.MethodOptions),
	"trace":   appAddMethod(fiber.MethodTrace),
	"patch":   appAddMethod(fiber.MethodPatch),
	"upgrade": wsAppUpgrade,
}

func appHandlerAsync(E *lue.Engine, app *Application, fn *lua.LFunction) fiber.Handler {
	return func(c *fiber.Ctx) error {
		E, _ := E.New()
		defer E.Close()
		env := E.Table(lua.EnvironIndex)
		if err := E.CallLFun(fn, env, 0, NewContext(E, app, c)); err != nil {
			return errWithStackTrace(err, c)
		}
		return nil
	}
}

func appSub(E *lue.Engine) int {
	app := E.Data(1).(*Application)
	prefix := E.String(2)
	subapp := new(Application)
	subapp.Router = app.Group(prefix)
	subapp.sub = true
	E.Push(E.Class(LTApplication, subapp))
	return 1
}

// appUse adds middleware to the app.
func appUse(E *lue.Engine) int {
	app := E.Data(1).(*Application)
	var values []any
	for i := 2; i <= E.Top(); i++ {
		switch val := E.Get(i).(type) {
		case lua.LString:
			values = append(values, string(val))
		case *lua.LTable:
			var list []string
			val.ForEach(func(_, v lua.LValue) {
				if str := lua.LVAsString(v); str != "" {
					list = append(list, str)
				}
			})
			values = append(values, list)
		case *lua.LFunction:
			values = append(values, appHandlerAsync(E, app, val))
		}
	}
	app.Use(values...)
	return 0
}

// appAdd adds a route to the app.
func appAdd(E *lue.Engine) int {
	app := E.Data(1).(*Application)
	method := E.String(2)
	path := E.String(3)
	handler := E.Fun(4)
	app.Add(method, path, appHandlerAsync(E, app, handler))
	return 0
}

// appAddMethod adds a route with specified method to the app.
func appAddMethod(method string) lue.Fun {
	return func(E *lue.Engine) int {
		app := E.Data(1).(*Application)
		path := E.String(2)
		handler := E.Fun(3)
		if method == methodAll {
			app.All(path, appHandlerAsync(E, app, handler))
		} else {
			app.Add(method, path, appHandlerAsync(E, app, handler))
		}
		return 0
	}
}

// appStart starts the app's listener.
func appStart(E *lue.Engine) int {
	app := E.Data(1).(*Application)
	if app.up {
		E.Error("app start: already started")
	}
	if app.sub {
		E.Error("app start: cannot start a subrouter")
	}
	listen := ""
	if E.Top() > 1 {
		listen = E.String(2)
	}
	app.up = true
	if err := app.c.Start(listen); err != nil {
		E.Error("app start: %v", err)
	}
	return 0
}

// appStart starts the app's listener.
func appReload(E *lue.Engine) int {
	app := E.Data(1).(*Application)
	if app.sub {
		E.Error("app reload: cannot reload a subrouter")
	}
	if app.c.Reload == nil {
		E.Error("app reload: not supported")
	}
	if err := app.c.Reload(); err != nil {
		E.Error("app reload: %v", err)
	}
	return 0
}

func appStop(E *lue.Engine) int {
	app := E.Data(1).(*Application)
	if !app.up {
		E.Error("app stop: already stopped")
	}
	if app.sub {
		E.Error("app stop: cannot stop a subrouter")
	}
	timeout := float64(0)
	if E.Top() > 1 {
		timeout = E.Number(2)
	}
	app.up = false
	const s = float64(time.Second)
	if err := app.c.Stop(time.Duration(timeout * s)); err != nil {
		E.Error("app stop: %v", err)
	}
	return 0
}

// errWithStackTrace adds a stack trace to the error if it's a Lua error.
func errWithStackTrace(e error, c *fiber.Ctx) error {
	if lerr, ok := e.(*lua.ApiError); ok {
		c.Locals("stacktrace", lerr.StackTrace)
		return errors.New(lerr.Object.String())
	}
	return e
}

File: types/application.lua
---@meta

--- Mirai 服务器
---@class Application : Router
local Application = {}

--- 启动 Mirai 服务器，开始监听新的连接。
function Application:start() end

--- 停止服务器。
---@param timeout number? 超时
function Application:stop(timeout) end

--- 设置传入处理器中的变量。
---
--- 请注意：值必须是一个只读变量，不能在运行的过程中改变。
---@param key string 变量名
---@param value any 值
function Application:set(key, value) end

--- 路由器
---@class Router
local Router = {}

--- 添加一个中间件。
---@param path string? 路径
---@param middleware fun(ctx: Context) 中间件
---@overload fun(self: Application, middleware: fun(ctx: Context))
function Router:use(path, middleware) end

--- 在指定的路径上，为指定的 HTTP 方法添加处理器。
---@param method string 方法
---@param path string 路径
---@param handler fun(ctx: Context) 处理器
function Router:add(method, path, handler) end

--- 在指定的路径上，为所有的 HTTP 方法添加处理器。
---@param path string 路径
---@param handler fun(ctx: Context)
function Router:all(path, handler) end

--- 在指定的路径上，为 GET 方法添加处理器。
---
--- GET 方法请求目标资源。这种请求不会改变数据状态。
---@param path string 路径
---@param handler fun(ctx: Context)
function Router:get(path, handler) end

--- 在指定的路径上，为 HEAD 方法添加处理器。
---
--- HEAD 方法请求目标资源的响应头，并且这些响应头与用 GET 方法请求得到的一致。服务器不能返回响应体。
---@param path string 路径
---@param handler fun(ctx: Context)
function Router:head(path, handler) end

--- 在指定的路径上，为 POST 方法添加处理器。
---
--- POST 方法向目标资源发送数据。这种请求通常会改变数据的状态。
---
--- PUT 和 POST 方法的区别是，PUT 方法是幂等的：调用一次与连续调用多次效果是相同的（即没有副作用），而连续调用多次相同的 POST 方法可能会产生副作用。
---@param path string 路径
---@param handler fun(ctx: Context)
function Router:post(path, handler) end

--- 在指定的路径上，为 PUT 方法添加处理器。
---
--- PUT 方法创建新的资源或用请求体替换目标资源。这种请求通常会改变数据的状态。
---
--- PUT 和 POST 方法的区别是，PUT 方法是幂等的：调用一次与连续调用多次效果是相同的（即没有副作用），而连续调用多次相同的 POST 方法可能会产生副作用。
---@param path string 路径
---@param handler fun(ctx: Context)
function Router:put(path, handler) end

--- 在指定的路径上，为 DELETE 方法添加处理器。
---
--- DELETE 方法用于删除目标资源。
---@param path string 路径
---@param handler fun(ctx: Context)
function Router:delete(path, handler) end

--- 在指定的路径上，为 DELETE 方法添加处理器。
---
--- CONNECT 方法可以开启与目标之间双向沟通的通道。可以用来创建隧道。
---@param path string 路径
---@param handler fun(ctx: Context)
function Router:connect(path, handler) end

--- 在指定的路径上，为 OPTIONS 方法添加处理器。
---
--- OPTIONS 方法请求目标资源的允许方法列表。客户端可以用这个方法指定一个 URL，或者用星号（*）来指代整个服务器。
---@param path string 路径
---@param handler fun(ctx: Context)
function Router:options(path, handler) end

--- 在指定的路径上，为 TRACE 方法添加处理器。
--- TRACE 方法请求沿着通往目标资源的路径进行信息回环测试。
---@param path string 路径
---@param handler fun(ctx: Context)
function Router:trace(path, handler) end

--- 在指定的路径上，为 TRACE 方法添加处理器。
--- PATCH 方法请求对资源进行部分修改。
---@param path string 路径
---@param handler fun(ctx: Context)
function Router:patch(path, handler) end

--- 在指定的路径上，为 WebSocket 协议升级添加处理器。
---@param path string 路径
---@param handler fun(ctx: Context)
function Router:upgrade(path, handler) end

app = setmetatable({}, { __index = Application })